engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/context/SuggestionBuilderTest.scala

```diff
@@ -854,7 +854,314 @@ class SuggestionBuilderTest extends AnyWordSpecLike with Matchers {
                 )
               ),
               selfType      = "Unnamed.Test",
-              returnType    = "Any",
+              returnType    = "Unnamed.Test.Value",
+              isStatic      = true,
+              documentation = None,
+              annotations   = Seq()
+            ),
+            Vector()
+          )
+        )
+      )
+    }
+
+    "build argument tag values from ascribed type" in {
+
+      val code =
+        """type Value
+          |    A
+          |    B
+          |
+          |type Auto
+          |
+          |foo (a : Value | Auto) = a
+          |""".stripMargin
+      val module = code.preprocessModule
+
+      build(code, module) shouldEqual Tree.Root(
+        Vector(
+          ModuleNode,
+          Tree.Node(
+            Suggestion.Type(
+              externalId    = None,
+              module        = "Unnamed.Test",
+              name          = "Value",
+              params        = Seq(),
+              returnType    = "Unnamed.Test.Value",
+              parentType    = Some(SuggestionBuilder.Any),
+              documentation = None
+            ),
+            Vector()
+          ),
+          Tree.Node(
+            Suggestion.Constructor(
+              externalId    = None,
+              module        = "Unnamed.Test",
+              name          = "A",
+              arguments     = Seq(),
+              returnType    = "Unnamed.Test.Value",
+              documentation = None,
+              annotations   = Seq()
+            ),
+            Vector()
+          ),
+          Tree.Node(
+            Suggestion.Constructor(
+              externalId    = None,
+              module        = "Unnamed.Test",
+              name          = "B",
+              arguments     = Seq(),
+              returnType    = "Unnamed.Test.Value",
+              documentation = None,
+              annotations   = Seq()
+            ),
+            Vector()
+          ),
+          Tree.Node(
+            Suggestion.Type(
+              externalId    = None,
+              module        = "Unnamed.Test",
+              name          = "Auto",
+              params        = Seq(),
+              returnType    = "Unnamed.Test.Auto",
+              parentType    = Some(SuggestionBuilder.Any),
+              documentation = None
+            ),
+            Vector()
+          ),
+          Tree.Node(
+            Suggestion.DefinedMethod(
+              externalId = None,
+              module     = "Unnamed.Test",
+              name       = "foo",
+              arguments = Seq(
+                Suggestion.Argument(
+                  "a",
+                  "Unnamed.Test.Value | Unnamed.Test.Auto",
+                  false,
+                  false,
+                  None,
+                  Some(Seq("..A", "..B", "..Auto"))
```

Comment:
@jdunkerley pointed out that there is a problem with offering `..A` in the `tagValues` - the IDE then doesn't know whether the `..A` has some constructor parameters or not!

@4e6 suggests to solve the problem by continuing to use FQN in `tagValues` array, but also add an information to each value, whether it can be _used in autoscoped form_ or not.

It would then be up to the IDE to decide whether to use _full qualified name_ or _autoscoped name_. How does that sound, @farmaazon?

Reply:
Actually, the FQNs in tag values are mostly used for deducing required import. Information about arguments is retrieved by Method Call. So just make sure the `..A` expression has proper Method Call in expressionUpdate. It may be that GUI still misinterprets `..` and that's why constructor arguments don't show already.

Reply:
> just make sure the `..A` expression has proper Method Call in `expressionUpdate`.

That's going to be a problem, Adam. As described by [autoscoped constructors docs](https://github.com/enso-org/enso/blob/a01aeab3a4acacf8f6a77ba94097f4cca16c8bc0/docs/syntax/types.md#type-definitions) the...

> argument `(..Just 5)` is recorded but not executed until it is send to the `describe` method.

... only later when ...

> ... The scope is ... known and the so far deferred `..` value gets evaluated.

E.g. there is **no Method Call** to report in `expressionUpdate` when the expression `(..Just 5)` is evaluated. 

**Brainstorming**:
- Only sometimes later a call happens
- However that must be similar to _suspended arguments_
- Does IDE handle suspended arguments well?
- If so, what do we need to do to send IDE enough information to also handle _autoscoped constructors_ well?

Reply:
> We don't have any logic for suspended arguments (`...`). I wasn't even aware something like that exists. So the answer is: probably not.

I was rather talking about `~` suspended arguments. However they seem to work in the IDE:

![image](https://github.com/enso-org/enso/assets/26887752/8a01f7cb-17b7-40e3-b2a3-19262a9117a7)

There is a method `confirm` that takes the `YesNo` type and `confirm2` that takes it suspended (evaluated only when needed). Still both the dropdowns look sane. The IDE knows arguments of `YesNo.Yes` as well as `YesNo.No` constructor.

Reply:
Constructor calls `Type.Cons` are wrapped in `FunctionCallInstrumentationNode` carrying information about the function call. This information is used to build a method pointer.
https://github.com/enso-org/enso/blob/d17d1ed4846aa27c9fe773be73f04278d4ae5119/engine/runtime-instrument-id-execution/src/main/java/org/enso/interpreter/instrument/IdExecutionInstrument.java#L228-L229

Reply:
> Constructor calls `Type.Cons` are wrapped in `FunctionCallInstrumentationNode` carrying information about the function call. This information is used to build a method pointer.

Great! We need to make sure that

https://github.com/enso-org/enso/blob/d17d1ed4846aa27c9fe773be73f04278d4ae5119/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/UnresolvedConstructor.java#L142

results in a call to `FunctionCallInstrumentationNode` then! I assume the call doesn't happen. Probably because we are missing instrumentation somewhere. Hopefully that's going to be fixable. I really believe there should be no difference between
```ruby
confirm2 (YesNo.No "because")
confirm (..No "because")
```
they should behave the same - e.g. defer the constructor creation until the usage of the argument.

Reply:
I think the issue is that the `UnresolvedConstructor` should be an `ExpressionNode` that implements `InstrumentableNode`.

Its execute method uses `InvokeFunctionNode` that inserts the `FunctionCallInstrumentationNode` correctly https://github.com/enso-org/enso/blob/d17d1ed4846aa27c9fe773be73f04278d4ae5119/engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/UnresolvedConstructor.java#L156-L157

But because the `UnresolvedConstructor` itself is not instrumentable, this call is not visible to the instrumentation

Reply:
Thank you Dmitry. `UnresolvedConstructor` is data, not code - e.g. it cannot be a `Node`. But otherwise great finding. Reported as
- #9381

Reply:
- #9381 - has just been integrated.